core/src/main/java/org/lflang/generator/c/CCmakeGenerator.java

```diff
@@ -136,6 +137,22 @@ CodeBuilder generateCMakeCode(
         cMakeCode.newLine();
         break;
       case RP2040:
+        // Attempt to set PICO_SDK_PATH if it is not already set.
```

Comment:
I'm not sure that this is the right approach. I don't think should be up to `lfc` to guess where the Pico SDK may be installed. The fact that an environment variable is used at all suggests that this it should be up to the user to specify it. I see that this code provides the convenience of doing that if the user forgets this, but it addresses only a single corner case, and it reinforces the idea that the output directory should contain copies of library code, which we're starting to realize has serious drawbacks. Is this all because we don't want to instruct users of the `rpi3+` template to invoke `nix develop`? Note that this also provides access to tools like `minicom`, which if you don't run `nix develop` will not be on your `$PATH`.
Perhaps we can do something more generic (and adopt the same approach for `lingo`), which is that if the package has a nix configuration, we automatically run `nix develop`, capture the environment variables, and use that environment for subsequent invocations of commands?

Reply:
`nix develop` will have to be run every time you start a new session. It launches a new shell. It does not configure the user environment.  Thus, with that approach, every lab exercise instructions will have to be modified to remind the user to run `nix develop` before they do anything else.

If you have a better approach, feel free to modify this and also modify the instructions in `lf-embedded-lab` (Installation.md, Tools.md, and, if necessary, each lab exercise).  I have spent way too much time already on configuration issues with this, and if I spend more time on them, the labs themselves have no chance of being ready in time for Fall semester.

Also, note that we are not using `minicom`, which doesn't work for me (it requires a mysterious "meta" key, which doesn't exist). We are using `screen` instead.

Reply:
> `nix develop` will have to be run every time you start a new session. It launches a new shell. It does not configure the user environment. I have spent way too much time already on configuration issues with this, and if I spend more time on them, the labs themselves have no chance of being ready in time for Fall semester.

Doesn't it configure the user environment _by launching a new shell?_
 
> If you have a better approach, feel free to modify this and also modify the instructions in `lf-embedded-lab` (Installation.md, Tools.md, and, if necessary, each lab exercise).

I plan to bring this up during the packaging meeting and will circle back. I'm happy to help with this!

> Also, note that we are not using `minicom`, which doesn't work for me (it requires a mysterious "meta" key, which doesn't exist). We are using `screen` instead.

Totally fine to use `screen` instead, we should just update the `nix` configuration to pull in that instead of `minicom` then. @gundralaa could you take care of this?

Reply:
> > `nix develop` will have to be run every time you start a new session. It launches a new shell. It does not configure the user environment. I have spent way too much time already on configuration issues with this, and if I spend more time on them, the labs themselves have no chance of being ready in time for Fall semester.
> 
> Doesn't it configure the user environment _by launching a new shell?_

In the new shell, you have environment variables set. If you log out and log back in, those environment variables are gone. You have to run `nix develop` again, and then do everything in the shell it launches.

> > Also, note that we are not using `minicom`, which doesn't work for me (it requires a mysterious "meta" key, which doesn't exist). We are using `screen` instead.
> 
> Totally fine to use `screen` instead, we should just update the `nix` configuration to pull in that instead of `minicom` then. @gundralaa could you take care of this?

I thought I did that already... ???

Reply:
My proposal is to add a target property called `pre-build`, which is like `build` but the provided command will be executed before building, and the environment of the pre-build command will be used when invoking the target tool chain (or build command). The `.lf` files in the rp3+ project would just specify `pre-build: nix develop` and they would just compile using `lfc`. I can go ahead and implement this if it sounds OK.

Reply:
I've thought about this, and I think that it is fair to assert using nix does require extra steps for the user but, in most cases, once you run nix develop, any terminal multiplexers or ides spawned from the nix shell inherit the environment variables. For example, spawning vscode using ``code .`` from a nix shell on ubuntu, which the students will most likely be using, keeps the applications bindings such that the vscode plugins use the nix versions of the packages like cmake.

The usage of nix for this project was to provide a way for users to easily get started without installing dependencies.  That being said, there aren't many dependencies and if someone doesn't want to use nix I think its very reasonable to direct them to setting the ``PICO_SDK_PATH`` manually in their own shell env and installing packages manually. To support this case, the setup documentation for the pico-platform needs to robust.

I personally don't think there is any further work that needs to be done here and adding new properties would only bloat the existing compiler. Use nix or dont it doesn't really matter, but lfc in my opinion should focus on code generation tasks instead of searching for dependencies.

Reply:
I generally agree with you, @gundralaa, which also motivated my initial comment. The idea behind `pre-build` is that it would also be useful for other external configurations, beyond `nix`, but I agree that it isn't a necessity but a "nice-to-have."

Reply:
Pre-build feels like an epicycle to me.  I would rather just use build, like we did with the nrf52.  And running nix every time also feels wrong to me.  It will slow down the compile, and you will likely get an ever growing number of nested shells running.

I really don't see what wrong with having lfc deal with what will be a common usage pattern that eliminates the need for an environment variable.  And it's easy to override by just setting the environment variable.

Reply:
> Pre-build feels like an epicycle to me. I would rather just use build, like we did with the nrf52. And running nix every time also feels wrong to me. It will slow down the compile, and you will likely get an ever growing number of nested shells running.

Running `nix develop` after the initial install runs really fast (.5 second on my machine). We can also just have `pre-build: configure-nix.sh`, which will be even faster. Either way, this won't lead to an ever growing number of nested shells (running a command in Java does not affect the shell from which `lfc` is invoked).

> I really don't see what wrong with having lfc deal with what will be a common usage pattern that eliminates the need for an environment variable. And it's easy to override by just setting the environment variable.

Dealing with these kinds of corner cases bloats the compiler. Such code will eventually not be used or break as common usage patterns change.

Reply:
More specifically, my concern is that the current fix deals with a single _corner case_ only, not with a _usage pattern_. The `pre-build` feature would be useful for a lot more than just setting this specific environment variable.

A build tool/package manager like `npm` also has user-specifiable commands like `install` or `pre-publish` to hook in stuff that needs to happen as part of building/packaging tasks that the tool handles, so I would argue that something like this is pretty common.

Reply:
Update on this: getting `pre-build` to work was easy, but not in combination with `nix develop`; indeed, because it spawns a new shell...

Reply:
It would allow one to execute something ahead of a normal build sequence that is not completely taken over by the user. If you use `build`, then the user is fully in charge of building (this is like running with `--no-compile` and handling the build steps externally).

As @gundralaa pointed out, there is no _need_ for any of this, but the point is provide a convenience that not just there to handle a single corner case but has broader utility.

Reply:
I would like to see the difference between a pre-build script and a build script. Is the diff more than one line?

Reply:
That depends on what's in the build script? My point is that there doesn't have to be a build script at all. I would say the `pre-build` is only useful when no `build` is specified. They can be mutually exclusive.

Reply:
I abandoned my changes because it turned out to be difficult to generalize this.